src/miniscript/satisfy.rs

```diff
@@ -535,11 +395,117 @@ impl_tuple_satisfier!(A, B, C, D, E, F);
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+/// Type of schnorr signature to produce
+pub enum SchnorrSigType {
+    /// Key spend signature
+    KeySpend {
+        /// Merkle root to tweak the key, if present
+        merkle_root: Option<TapNodeHash>,
+    },
+    /// Script spend signature
+    ScriptSpend {
+        /// Leaf hash of the script
+        leaf_hash: TapLeafHash,
+    },
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+/// Placeholder for some data in a [`Plan`]
+///
+/// [`Plan`]: crate::plan::Plan
+pub enum Placeholder<Pk: MiniscriptKey> {
+    /// Public key and its size
+    Pubkey(Pk, usize),
+    /// ECDSA signature given the raw pubkey
+    EcdsaSigPk(Pk),
+    /// Schnorr signature and its size
+    SchnorrSig(Pk, SchnorrSigType, usize),
+    /// SHA-256 preimage
+    Sha256Preimage(Pk::Sha256),
+    /// HASH256 preimage
+    Hash256Preimage(Pk::Hash256),
+    /// RIPEMD160 preimage
+    Ripemd160Preimage(Pk::Ripemd160),
+    /// HASH160 preimage
+    Hash160Preimage(Pk::Hash160),
+    /// Hash dissatisfaction (32 bytes of 0x00)
+    HashDissatisfaction,
+    /// OP_1
+    PushOne,
+    /// \<empty item\>
+    PushZero,
+    /// Taproot leaf script
+    TapScript(ScriptBuf),
+    /// Taproot control block
+    TapControlBlock(ControlBlock),
+}
+
+impl<Pk: MiniscriptKey> fmt::Display for Placeholder<Pk> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use Placeholder::*;
+        match self {
+            Pubkey(pk, size) => write!(f, "Pubkey(pk: {}, size: {})", pk, size),
+            EcdsaSigPk(pk) => write!(f, "EcdsaSigPk(pk: {})", pk),
+            SchnorrSig(pk, tap_leaf_hash, size) => write!(
+                f,
+                "SchnorrSig(pk: {}, tap_leaf_hash: {:?}, size: {})",
+                pk, tap_leaf_hash, size
+            ),
+            Sha256Preimage(hash) => write!(f, "Sha256Preimage(hash: {})", hash),
+            Hash256Preimage(hash) => write!(f, "Hash256Preimage(hash: {})", hash),
+            Ripemd160Preimage(hash) => write!(f, "Ripemd160Preimage(hash: {})", hash),
+            Hash160Preimage(hash) => write!(f, "Hash160Preimage(hash: {})", hash),
+            HashDissatisfaction => write!(f, "HashDissatisfaction"),
+            PushOne => write!(f, "PushOne"),
+            PushZero => write!(f, "PushZero"),
+            TapScript(script) => write!(f, "TapScript(script: {})", script),
+            TapControlBlock(control_block) => write!(
+                f,
+                "TapControlBlock(control_block: {})",
+                bitcoin::consensus::encode::serialize_hex(&control_block.serialize())
+            ),
+        }
+    }
+}
+
+impl<Pk: MiniscriptKey + ToPublicKey> Placeholder<Pk> {
+    /// Replaces the placeholders with the information given by the satisfier
+    pub fn satisfy_self<Sat: Satisfier<Pk>>(&self, sat: &Sat) -> Option<Vec<u8>> {
+        match self {
+            Placeholder::Pubkey(pk, _) => Some(pk.to_public_key().to_bytes()),
```

Comment:
This would work in case of ECDSA Sig. But with the Schnorr sig, It is not fulfilling. 
`Placeholder::Pubkey(pk, _)`. Here the second argument is length so we can have a length instead of `_` and check if it was 32 or something else. If 32 then do `Some(pk.to_x_only_pubkey().serialize().to_vec())`

Reply:
Thanks! I think it should be: if the size is 33, then I have to use to_x_only_pubkey, otherwise I can use to_public_key. This is because we push Ctx::pk_len, which is 33 for Tap. Does this make sense?
```rust
            Placeholder::Pubkey(pk, size) => {
                if *size == 33 {
                    Some(pk.to_x_only_pubkey().serialize().to_vec())
                } else {
                    Some(pk.to_public_key().to_bytes())
                }
            }
```